packages/wallet/api/test/test-wallet-marshal.js

```diff
@@ -0,0 +1,166 @@
+  });
+
+  /** @type {Brand[]} */
+  const [b1] = context.fromBoard.unserialize(ammMetricsCapData);
+  /** @type {Brand[]} */
+  const [b2] = context.fromBoard.unserialize(amm.getMetrics());
+  t.is(b1, b2, 'unserialization twice from same source');
+
+  const myWallet = makeOnChainWallet(board);
+  myWallet.suggestIssuer('ATOM', 'board011');
+  const walletCapData = myWallet.publish();
+  t.deepEqual(walletCapData, {
+    body: JSON.stringify([
+      {
+        brand: { '@qclass': 'slot', iface: 'Alleged: ATOM brand', index: 0 },
+        brandPetname: 'ATOM',
+        currentAmount: { brand: { '@qclass': 'slot', index: 0 }, value: 100 },
+        meta: { id: 0 },
+        purse: { '@qclass': 'slot', iface: 'Alleged: ATOM purse', index: 1 },
+        pursePetname: 'ATOM purse',
+      },
+    ]),
+    slots: ['board011', 'purse:ATOM'],
+  });
+
+  const walletState = context.fromMyWallet.unserialize(walletCapData);
+  t.is(walletState[0].brand, b1, 'unserialization across sources');
+
+  t.throws(
+    () => context.fromBoard.unserialize(walletCapData),
+    { message: /bad board slot/ },
+    'AMM cannot refer to purses',
+  );
+
+  t.throws(
+    () => context.fromMyWallet.serialize(Far('widget', {})),
+    {
+      message: /unregistered/,
+    },
+    'import context never exports unregistered objects',
+  );
+});
+
+test('ensureBoardId allows re-registration; initBoardId does not', t => {
+  const brandM = Far('Moola brand', {});
+  const brandS = Far('Semolean brand', {});
+
+  const context = makeExportContext();
+  context.initBoardId('board1', brandM);
+  t.throws(() => context.initBoardId('board1', brandM));
+  context.ensureBoardId('board1', brandM);
+  t.throws(() => context.ensureBoardId('board12', brandM));
+  context.ensureBoardId('board12', brandS);
+  t.throws(() => context.initBoardId('board12', brandM));
+});
+
+test('makeExportContext.serialize handles unregistered identites', t => {
+  const brand = Far('Zoe invitation brand', {});
+  const instance = Far('amm instance', {});
+  const invitationAmount = harden({ brand, value: [{ instance }] });
+
+  const context = makeExportContext();
+  context.initBoardId('board1', brand);
+  const actual = context.serialize(invitationAmount);
+
+  t.deepEqual(actual, {
+    body: JSON.stringify({
+      brand: {
+        '@qclass': 'slot',
+        iface: 'Alleged: Zoe invitation brand',
+        index: 0,
+      },
+      value: [
+        {
+          instance: {
+            '@qclass': 'slot',
+            iface: 'Alleged: amm instance',
+            index: 1,
+          },
+        },
+      ],
+    }),
+    slots: ['board1', 'unknown:1'],
+  });
+
+  t.deepEqual(context.unserialize(actual), invitationAmount);
+
+  const myPayment = /** @type {Payment} */ (
+    Far('payment', { getAllegedBrand: () => assert.fail('no impl') })
+  );
+  context.savePaymentActions(myPayment);
+  const cap2 = context.serialize(myPayment);
+  t.deepEqual(cap2, {
+    body: JSON.stringify({
```

Comment:
when these don't match, it'll be a string diff.  consider comparing the object with `JSON.parse(cap2)`.

If not comparing objects then it should be `t.is`

Reply:
The test is supposed to be at the CapData level, where `body` is a string. Writing the expected `body` using `JSON.stringify()` makes it easier to read. I suppose I could add a comment...

Meanwhile, your comment prompts me to think about this a bit more carefully... In the API, `slots` are compared by identity, so `t.deepEqual()` is only correct because we happen to be using primitives (strings). In nearby designs, the slots are `{ kind: 'purse', id: 123}` objects where it's really easy to have things that pass `t.deepEqual()` but not `t.is()`.